File: ext/doc/Dse/Keyspace.php
<?php

/*
  Copyright (c) 2016 DataStax, Inc.
  This software can be used solely with DataStax Enterprise. Please consult the
  license at http://www.datastax.com/terms/datastax-dse-driver-license-terms
*/

namespace Dse;

/**
 * A PHP representation of a keyspace
 */
interface Keyspace {

    /**
     * Returns keyspace name
     *
     * @return string Name
     */
    public function name();

    /**
     * Returns replication class name
     *
     * @return string Replication class
     */
    public function replicationClassName();

    /**
     * Returns replication options
     *
     * @return Map Replication options
     */
    public function replicationOptions();

    /**
     * Returns whether the keyspace has durable writes enabled
     *
     * @return string Whether durable writes are enabled
     */
    public function hasDurableWrites();

    /**
     * Returns a table by name
     *
     * @param string $name Table name
     *
     * @return Table|null Table instance or null
     */
    public function table($name);

    /**
     * Returns all tables defined in this keyspace
     *
     * @return array An array of `Table` instances
     */
    public function tables();

    /**
     * Get user type by name
     *
     * @param string $name User type name
     *
     * @return Type\UserType|null A user type or null
     */
    public function userType($name);

    /**
     * Get all user types
     *
     * @return array An array of user types
     */
    public function userTypes();

    /**
     * Get materialized view by name
     *
     * @param string $name Materialized view name
     *
     * @return MaterizedView|null A materialized view or null
     */
    public function materializedView($name);

    /**
     * Gets all materialized views
     *
     * @return array An array of materialized views
     */
    public function materializedViews();

    /**
     * Get a function by name and signature
     *
     * @param string $name Function name
     * @param string|Type $params Function arguments
     *
     * @return Function|null A function or null
     */
    public function function_($name, ...$params);

    /**
     * Get all functions
     *
     * @return array An array of functions
     */
    public function functions();

    /**
     * Get an aggregate by name and signature
     *
     * @param string $name Aggregate name
     * @param string|Type $params Aggregate arguments
     *
     * @return Aggregate|null An aggregate or null
     */
    public function aggregate($name, ...$params);

    /**
     * Get all aggregates
     *
     * @return array An array of aggregates
     */
    public function aggregates();

}

File: ext/src/Keyspace.c
/*
  Copyright (c) 2017 DataStax, Inc.

  This software can be used solely with DataStax Enterprise. Please consult the
  license at http://www.datastax.com/terms/datastax-dse-driver-license-terms
*/

#include "php_driver.h"

zend_class_entry *php_driver_keyspace_ce = NULL;

ZEND_BEGIN_ARG_INFO_EX(arginfo_name, 0, ZEND_RETURN_VALUE, 1)
  ZEND_ARG_INFO(0, name)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_signature, 0, ZEND_RETURN_VALUE, 1)
  ZEND_ARG_INFO(0, name)
  ZEND_ARG_INFO(0, ...)
ZEND_END_ARG_INFO()

ZEND_BEGIN_ARG_INFO_EX(arginfo_none, 0, ZEND_RETURN_VALUE, 0)
ZEND_END_ARG_INFO()

static zend_function_entry php_driver_keyspace_methods[] = {
  PHP_ABSTRACT_ME(Keyspace, name, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, replicationClassName, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, replicationOptions, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, hasDurableWrites, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, table, arginfo_name)
  PHP_ABSTRACT_ME(Keyspace, tables, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, userType, arginfo_name)
  PHP_ABSTRACT_ME(Keyspace, userTypes, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, materializedView, arginfo_name)
  PHP_ABSTRACT_ME(Keyspace, materializedViews, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, function, arginfo_signature)
  PHP_ABSTRACT_ME(Keyspace, functions, arginfo_none)
  PHP_ABSTRACT_ME(Keyspace, aggregate, arginfo_signature)
  PHP_ABSTRACT_ME(Keyspace, aggregates, arginfo_none)
  PHP_FE_END
};

void php_driver_define_Keyspace(TSRMLS_D)
{
  zend_class_entry ce;

  INIT_CLASS_ENTRY(ce, PHP_DRIVER_NAMESPACE "\\Keyspace", php_driver_keyspace_methods);
  php_driver_keyspace_ce = zend_register_internal_class(&ce TSRMLS_CC);
  php_driver_keyspace_ce->ce_flags |= ZEND_ACC_INTERFACE;
}
